### ISSUE TITLE
Wrong reordering of match arms with ranges

### ISSUE THREAD

**Issue by lifthrasiir** [rust-lang/rust#18060]
Adopted and minimized from the HTTP method parsing routine of [hyperium/hyper](https://github.com/hyperium/hyper):

``` rust
println!("{}", match (1u, 3u) { (0, 2...5) => 1, (1, 3) => 2, (_, 2...5) => 3, (_, _) => 4u });
println!("{}", match (1u, 3u) {                  (1, 3) => 2, (_, 2...5) => 3, (_, _) => 4u });
println!("{}", match (1u, 7u) { (0, 2...5) => 1, (1, 7) => 2, (_, 2...5) => 3, (_, _) => 4u });
```

All three lines should print 2, but the first line prints 3 on the current nightly (`rustc 0.13.0-nightly (40b244973 2014-10-14 23:22:20 +0000)`). The `match` trans apparantly groupped the second pattern even when it may result in the wrong result.

**Comment by remram44** [comment]
Still happens :(

**Comment by niconii** [comment]
This seems to be fixed as of beta?

**Comment by solson** [comment]
Yes, this is already fixed in beta 1.12: https://is.gd/fwGJ30
